subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/ZipHasher.java

```diff
@@ -32,82 +34,99 @@
 import org.slf4j.LoggerFactory;
 
 import javax.annotation.Nullable;
+import java.io.File;
 import java.io.IOException;
-import java.io.InputStream;
-import java.nio.file.Files;
-import java.nio.file.Paths;
 import java.util.List;
-import java.util.zip.ZipEntry;
-import java.util.zip.ZipInputStream;
+import java.util.Set;
 
-public class JarHasher implements RegularFileHasher, ConfigurableNormalizer {
-    private static final Logger LOGGER = LoggerFactory.getLogger(JarHasher.class);
+public class ZipHasher implements RegularFileHasher, ConfigurableNormalizer {
 
-    private final ResourceHasher classpathResourceHasher;
-    private final ResourceFilter classpathResourceFilter;
+    private static final Set<String> KNOWN_ZIP_EXTENSIONS = ImmutableSet.of("zip", "jar", "war", "ear", "apk");
+    private static final Logger LOGGER = LoggerFactory.getLogger(ZipHasher.class);
 
-    public JarHasher(ResourceHasher classpathResourceHasher, ResourceFilter classpathResourceFilter) {
-        this.classpathResourceHasher = classpathResourceHasher;
-        this.classpathResourceFilter = classpathResourceFilter;
+    public static boolean isZipFile(final String name) {
```

Comment:
I'm wondering if there's a better way to do this.

First, `hasExtensionIgnoresCase` will match e.g. `clear` as being a zip file since the trailing dot is not matched for some reason.

Second, currently we're iterating over a (admittedly very small) collection and checking lowercase equality.

If we already had the extension, this could be simply:
```java
return KNOWN_ZIP_EXTENSIONS.contains(extension);
```
Which would be simpler and a lot faster.
If we don't have the extension, we could still do
```java
String extension = Files.getFileExtension(fileName).toLowerCase()
return KNOWN_ZIP_EXTENSIONS.contains(extension);
```
Which would still be simple and basically have [the same speed](https://gist.github.com/paplorinc/e81971ca19c6a5ca5f90ac435ba48373).

What do you think?

Reply:
This couldn't leave me alone, seeing that all files are going through this filer. I know it's an insignificant part of the process, but the optimizations aren't complicated either.
I've tried `TreeSet`, Pattern matching and ended up doing a custom parser and I think it's worth considering it here:
![image](https://user-images.githubusercontent.com/1841944/57160640-a5530600-6de9-11e9-84e4-5114557dadce.png)
i.e. with a custom parser it would be 23x faster on average to partition files into the two groups (inner-zip vs normal file).

Reply:
Let's not optimize where there is no problem. This method doesn't show up in CPU graphs, not even 0.1%.

Reply:
Without a provided benchmark I had to guess.
If you say it's not a hot method, I agree, let's not over-optimize.
I think my simplification comment still applies, i.e.
```java
public static boolean isZipFile(String fileName) {
    String extension = Files.getFileExtension(fileName);
    return KNOWN_ZIP_EXTENSIONS.contains(extension);
}
```
Do you agree, @oehme?